### PR TITLE
[graphql/rpc] Graphql impl query complexity

### PR DESCRIPTION
**Pull request by oxade**
## Description 

Initial logic for graphql query and depth based limiting

## Test Plan 

Manual

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

**Review by amnn — COMMENTED** [review]
The depth limit stuff looks great, but I have some concerns on the direction of query complexity.  In the limit of the current approach we will have to tune per-query parameters to correlate with the results you would get from `EXPLAIN`, which is not a worthwhile goal.

The benefit is that we would be able to do this estimation entirely during validation, but the downside is that I don't think we will ever accurately model `EXPLAIN`'s output with the tools that async-graphql gives us for query complexity, so we will forever be costing queries with two similar but subtly different models, which will lead to a poor UX (In fact, I thought that we ended up deciding we couldn't use async-graphql's query complexity system because it was too limiting).

All this being said, there is one gap in our limits story that something like (but not exactly the same as) query complexity could solve, which is to limit the number of database round-trips:

```graphql
query {
  chainIdentifier
  objectConnection(filter: { objectIds: ["0x1", "0x2", "0x3"] }) {
    edges {
      node {
        previousTransactionBlock {
          effects { digest }
          gasInput { 
            gasSponsor { location }
            gasPayment { owner }
          }
        }
      }
    }
  }
}
```

In the best case, using data loaders, we're going to need to perform four database round trips for this example: 
- One for the chain identifier, 
- One to get objects,
- One to get the transaction blocks of those objects.  
- One to get the gas payment objects for the transaction blocks.

And the following sub-queries be treated as just projects:
- `effects { digest }` is part of a field from the `transactions` table, as is `gasInput`.
- `gasSponsor { location }` is "free" because `location` is part of the primary key of the `addresses` table, so you don't need to load an address to get it, it's already part of the field in `gasInput`.
 
Query depth only approximates this, but we could discover the true round trip number using a validation pass.

This requires knowing which queries are projections (extracting fields from an already loaded structure), which fields are primary keys, which queries are true edges (a foreign key that needs to be loaded in another table), and in that case which table it's loaded from.

Algorithm works top-down, by building a tree of table loads from the query AST, which would look like this:

```
- <chain identifier>
- <objects>
   - <transactions>
      - <objects> 
```

A node in the query AST gets a node in the table AST if it's a "true edge" and the sub-query it points to includes anything that isn't a primary key field.  The complexity is the number of nodes in the table tree.

It's unclear to me whether the data loader pattern is clever enough to merge "sibling" loads to the same database (i.e. if I had two `objectConnection` queries next to each other), but if it is then you would need to take that into account by merging siblings with matching table names in the tree (I think it probably doesn't do this though).

This whole process is calculated similarly to query complexity, it's just the details of the recursion that are different, so I think it would have to be implemented as an [Extension](https://async-graphql.github.io/async-graphql/en/extensions_inner_working.html#validation) that adds validation middleware.

Let me know what you think.

**Comment by wlmyng** [comment]
> 

iiuc from this doc page https://async-graphql.github.io/async-graphql/en/dataloader.html and example https://github.com/async-graphql/examples/blob/master/tide/dataloader-postgres/src/main.rs I think DataLoaders would indeed merge "sibling" loads to the same database

**Comment by wlmyng** [comment]
> It's unclear to me whether the data loader pattern is clever enough to merge "sibling" loads to the same database (i.e. if I had two `objectConnection` queries next to each other), but if it is then you would need to take that into account by merging siblings with matching table names in the tree (I think it probably doesn't do this though).

iiuc from this doc page https://async-graphql.github.io/async-graphql/en/dataloader.html and example https://github.com/async-graphql/examples/blob/master/tide/dataloader-postgres/src/main.rs I think DataLoaders would indeed merge "sibling" loads to the same database

**Comment by amnn** [comment]
@wlmyng, from my reading of those docs, `DataLoader`s do some amount of merging -- e.g. if you load the todos of all users, it will issue one read for all the todos, even though each is for a separate user, and in that sense it is merging siblings, but what I couldn't figure out from the docs was whether in a case like:

```graphql
query {
  t1: todo(T1) { user }
  t2: todo(T2) { user }
}
```

whether data loaders would perform 2, 3 or 4 reads.  A very smart implementation could get it down to two, but then it gets more complicated in a case like:

```graphql
query {
  t1: todos(before: "2023-09-07") { user }
  t2: todos(priority: 1) { user }  
}
```

Now the top-level queries for TODOs are using different filters, so does this take 3 or 4 round-trips, and if the first case can be done with 2 reads, how does the pattern reliably detect when it's in that case vs this case?

So my assumption is that the pattern is making the minimal possible assumption, and it's only merging non-syntactic sibling reads (where it's not a sibling in the query, but because one query produced multiple results, and you want to load more data for each of those results).

File: crates/sui-graphql-rpc/src/limits/complexity.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+}
+
+impl Default for ComplexityConfigEntry {
+    fn default() -> Self {
+        Self {
+            base_complexity: DEFAULT_BASE_COMPLEXITY,
+            child_complexity_multiplier: DEFAULT_CHILD_COMPLEXITY_MULTIPLIER,
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-graphql-rpc/src/limits/complexity.rs:31]: changed them to optional

File: crates/sui-graphql-rpc/src/limits/complexity.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+/// Standard calculation for complexity
+pub(crate) fn standard_calc(
+    c: &ComplexityConfigEntry,
+    first: Option<u64>,
+    last: Option<u64>,
+    child_complexity: usize,
+) -> usize {
+    let num_nodes = max(first.unwrap_or(0) + last.unwrap_or(0), 1) as usize;
```
[the remark this answers is not in the document]
Comment (oxade) [reply crates/sui-graphql-rpc/src/limits/complexity.rs]: indeed; fixed

**Comment by oxade** [comment]
> The depth limit stuff looks great, but I have some concerns on the direction of query complexity. In the limit of the current approach we will have to tune per-query parameters to correlate with the results you would get from `EXPLAIN`, which is not a worthwhile goal.

I should probably have specified; the changes in this PR do not negate the need for custom `EXPLAIN` based limiting or more advanced validation-based limiting.
The changes here were just the first step moving us in some direction.
Regarding the broader more intricate limiting algorithms you specified, you're correct and the interaction with the data loader complicates things. I plan to design this incrementally and we can try different extensions and see what works.

**Comment by oxade** [comment]
We spoke internally and decided to keep it simple by just limiting the number of nodes for now.
Changes made in superceding [PR](https://github.com/MystenLabs/sui/pull/13855)

**Pull request closed without merging**